File: gosrc/ir/ir.go
package ir

import "fmt"

type Type interface {
	String() string
}
type TypeName struct {
	Inner string
}

func (t *TypeName) String() string {
	return t.Inner
}

type TypePointer struct {
	InnerType Type
}

func (t *TypePointer) String() string {
	return fmt.Sprintf("*%s", t.InnerType)
}

type TypeArray struct {
	InnerType Type
}

func (t *TypeArray) String() string {
	return fmt.Sprintf("[]%s", t.InnerType)
}

type Package struct {
	Modules []Module
}

type Module struct {
	Funcs []*Func
}

func (m *Module) AddFunc(f *Func) {
	m.Funcs = append(m.Funcs, f)
}

type Struct struct {
}

type Impl struct {
}

type Func struct {
	Name       string
	Exported   bool
	ReturnType Type
	Block      *Block
	BlockCount int
}

func (f *Func) Print() {
	fmt.Printf("Func %s:\n", f.Name)
	f.Block.Print(0)
}

func (f *Func) NewBlock(parent *Block) *Block {
	b := &Block{
		Number: f.BlockCount,
		F:      f,
		Parent: parent,
		Insts:  make([]Inst, 0),
	}
	f.BlockCount += 1
	return b
}

type Variable struct {
	Name       string
	Type       Type
	Block      *Block
	InsertedAt Ref
	IsParam    bool
	IsConst    bool
}

type Ref = int

type Block struct {
	Number    int
	Parent    *Block
	F         *Func
	Variables []*Variable
	Insts     []Inst
}

func (b *Block) Print(level int) {
	printf(level, "Block %d:\n", b.Number)
	for i, v := range b.Insts {
		printf(level, "\t%d:", i)
		v.Print(level + 1)
	}
}

func (b *Block) findVariable(name string) *Variable {
	s := b
	for s != nil {
		for _, i := range s.Variables {
			if name == i.Name {
				return i
			}
		}
		s = s.Parent
	}
	return nil
}

func (b *Block) Append(i Inst) Ref {
	loc := len(b.Insts)
	b.Insts = append(b.Insts, i)
	return loc
}

func (b *Block) Ret() Ref {
	return b.Append(&InstRet{})
}

func (b *Block) RetValue(value Ref) Ref {
	return b.Append(&InstRet{
		HasValue: true,
		Value:    value,
	})
}

func (b *Block) Add(lhs Ref, rhs Ref) Ref {
	return b.Append(&InstAdd{
		Rhs: rhs,
		Lhs: lhs,
	})
}

func (b *Block) Eq(lhs Ref, rhs Ref) Ref {
	return b.Append(&InstEq{
		Rhs: rhs,
		Lhs: lhs,
	})
}

func (b *Block) NotEq(lhs Ref, rhs Ref) Ref {
	return b.Append(&InstNotEq{
		Rhs: rhs,
		Lhs: lhs,
	})
}

func (b *Block) Int(value int) Ref {
	return b.Append(&InstInt{
		Value: value,
	})
}

func (b *Block) If(condition Ref, then *Block, _else *Block) Ref {
	return b.Append(&InstIf{
		Condition: condition,
		Then:      then,
		Else:      _else,
	})
}

func (b *Block) Var(name string, _type Type, isConst bool, equals Ref) Ref {
	if b.findVariable(name) != nil {
		// TODO(caleb): compiler error
		panic("Variable already exists")
	}
	v := &Variable{
		Name:       name,
		Type:       _type,
		Block:      b,
		InsertedAt: len(b.Insts),
		IsParam:    false,
		IsConst:    isConst,
	}
	b.Variables = append(b.Variables, v)
	return b.Append(&InstVar{
		V:      v,
		Equals: equals,
	})
}

func (b *Block) Load(name string) Ref {
	v := b.findVariable(name)
	if v == nil {
		// TODO(caleb): compiler error
		panic("Could not find variable")
	}
	return b.Append(&InstLoad{
		V: v,
	})
}

func (b *Block) Store(name string, equals Ref) Ref {
	v := b.findVariable(name)
	if v == nil {
		// TODO(caleb): compiler error
		panic("Could not find variable")
	}
	return b.Append(&InstStore{
		V:      v,
		Equals: equals,
	})
}

func (b *Block) Call(name string, params []Ref) Ref {
	return b.Append(&InstCall{
		Name:   name,
		Params: params,
	})
}

type Inst interface {
	Print(level int)
}

// Represents a block inside another block
type InstBlock struct {
	Block *Block
}

func (i *InstBlock) Print(level int) {
	fmt.Printf("Block:\n")
	i.Block.Print(level + 1)
}

type InstAdd struct {
	Lhs Ref
	Rhs Ref
}

func (i *InstAdd) Print(level int) {
	fmt.Printf("Add %d %d\n", i.Lhs, i.Rhs)
}

type InstEq struct {
	Lhs Ref
	Rhs Ref
}

func (i *InstEq) Print(level int) {
	fmt.Printf("Eq %d %d\n", i.Lhs, i.Rhs)
}

type InstNotEq struct {
	Lhs Ref
	Rhs Ref
}

func (i *InstNotEq) Print(level int) {
	fmt.Printf("NotEq %d %d\n", i.Lhs, i.Rhs)
}

type InstRet struct {
	HasValue bool
	Value    Ref
}

func (i *InstRet) Print(level int) {
	if i.HasValue {
		fmt.Printf("Ret %d\n", i.Value)
	} else {
		fmt.Printf("Ret\n")
	}
}

type InstInt struct {
	Value int
}

func (i *InstInt) Print(level int) {
	fmt.Printf("Int %d\n", i.Value)
}

type InstIf struct {
	Condition Ref
	Then      *Block
	Else      *Block
}

func (i *InstIf) Print(level int) {
	fmt.Printf("If %d\n", i.Condition)

	i.Then.Print(level + 1)
	if i.Else != nil {
		i.Else.Print(level + 1)
	}
}

type InstVar struct {
	V      *Variable
	Equals Ref
}

func (i *InstVar) Print(level int) {
	fmt.Printf("Var (%s: %s) %d\n", i.V.Name, i.V.Type, i.Equals)
}

type InstLoad struct {
	V *Variable
}

func (i *InstLoad) Print(level int) {
	fmt.Printf("Load (%s: %s)\n", i.V.Name, i.V.Type)
}

type InstStore struct {
	V      *Variable
	Equals Ref
}

func (i *InstStore) Print(level int) {
	fmt.Printf("Store (%s: %s) %d\n", i.V.Name, i.V.Type, i.Equals)
}

type InstCall struct {
	Name   string
	Params []Ref
}

func (i *InstCall) Print(level int) {
	fmt.Printf("Call %s(n%d)\n", i.Name, len(i.Params))
}

func printf(level int, format string, a ...any) {
	for i := 0; i < level; i++ {
		fmt.Print("\t")
	}
	fmt.Printf(format, a...)
}

File: src/ir.cpp
#include "ir.h"

ref<Block> Block::new_child_block() {
    auto block = make_ref<Block>();
    block->number = func->block_count++;
    block->func = func;
    block->parent = this;
    return block;
}

ref<Variable> Block::find_variable(const std::string& name) {
    auto block = this;
    while (block != nullptr) {
        for(auto var: block->vars) {
            if (var->name == name) {
                return var;
            }
        }
        block = block->parent;
    }
    return nullptr;
}

Ref Block::add_inst(InstType type, Data data) {
    Ref r = insts.size();
    insts.push_back(Inst {
        type,
        data
    });
    return r;
}

Ref Block::block(ref<Block> block) {
    return add_inst(InstBlock, DataBlock{
        block
    });
}

Ref Block::call(std::string name, std::vector<Ref> params) {
    return add_inst(InstCall, DataCall{
        name,
        params,
    });
}

Ref Block::ret() {
    return add_inst(InstRet, DataNull{});
}

Ref Block::retval(Ref value) {
    return add_inst(InstRetVal, DataUnary{
        value
    });
}

Ref Block::if_(Ref condition, ref<Block> then, std::optional<ref<Block>> else_) {
    return add_inst(InstIf, DataIf{
        condition,
        then,
        else_
    });
}

Ref Block::var(const std::string& name, Type type, bool is_const, Ref equals) {
    if (find_variable(name) != nullptr) {
        // TODO(caleb): compiler error
    }
    auto var = make_ref<Variable>();
    var->name = name;
    var->type = type;
    var->is_const = is_const;
    var->number = func->var_count++;
    vars.push_back(var);
    return add_inst(InstVar, DataVar{
        var,
        equals
    });
}

Ref Block::store(const std::string& name, Ref value) {
    auto var = find_variable(name);
    if (var == nullptr) {
        // TODO(caleb): compiler error
    }
    return add_inst(InstStore, DataStore{
        var,
        value
    });
}

Ref Block::load(std::string name) {
    auto var = find_variable(name);
    if (var == nullptr) {
        // TODO(caleb): compiler error
    }
    return add_inst(InstLoad, DataLoad{
        var
    });
}

Ref Block::add(Ref lhs, Ref rhs) {
    return add_inst(InstAdd, DataBinOp{
        lhs,
        rhs
    });
}

Ref Block::eq(Ref lhs, Ref rhs) {
    return add_inst(InstEq, DataBinOp{
        lhs,
        rhs
    });
}

Ref Block::noteq(Ref lhs, Ref rhs) {
    return add_inst(InstNotEq, DataBinOp{
        lhs,
        rhs
    });
}

Ref Block::int_(uint64_t value) {
    return add_inst(InstInt, DataInt{
        value
    });    
}

void print_indent(uint32_t indent) {
    for(uint32_t i = 0; i < indent; i++) {
        printf("  ");
    }
}

void print_block(ref<Block> b, uint32_t indent) {
    print_indent(indent);
    printf("Block %lld\n", b->number);
    for(uint32_t i = 0; i < b->insts.size(); i++) {
        Inst& inst = b->insts[i];
        print_indent(indent);
        printf("%d: ", i);
        switch(inst.type) {
        case InstBlock: {
            printf("Block: \n");
            auto data = std::get<DataBlock>(inst.data);
            print_block(data.block, indent + 1);
            break;
        }
        case InstCall: {
            auto data = std::get<DataCall>(inst.data);
            printf("Call %s (n %lld)\n", data.name.c_str(), data.params.size());
            break;
        }
        case InstRetVal: {
            auto data = std::get<DataUnary>(inst.data);
            printf("Ret %lld\n", data.value);
            break;
        }
        case InstRet: {
            printf("Ret\n");
            break;
        }
        case InstIf: {
            auto data = std::get<DataIf>(inst.data);
            printf("If %lld\n", data.condition);
            print_indent(indent);
            printf("then:\n");
            print_block(data.then, indent + 1);
            if (auto else_ = data.else_) {
                print_indent(indent);
                printf("else:\n");
                print_block(*else_, indent + 1);
            }
            break;
        }
        case InstVar: {
            auto data = std::get<DataVar>(inst.data);
            printf("Var %s: %s = %lld\n", data.var->name.c_str(), "unknown", data.equals);
            break;
        }
        case InstStore: {
            auto data = std::get<DataStore>(inst.data);
            printf("Store %s: %s = %lld\n", data.var->name.c_str(), "unknown", data.equals);
            break;
        }
        case InstLoad: {
            auto data = std::get<DataLoad>(inst.data);
            printf("Load %s: %s\n", data.var->name.c_str(), "unknown");
            break;
        }
        case InstAdd: {
            auto data = std::get<DataBinOp>(inst.data);
            printf("Add %lld %lld\n", data.rhs, data.lhs);
            break;
        }
        case InstEq: {
            auto data = std::get<DataBinOp>(inst.data);
            printf("Eq %lld %lld\n", data.rhs, data.lhs);
            break;
        }
        case InstNotEq: {
            auto data = std::get<DataBinOp>(inst.data);
            printf("NotEq %lld %lld\n", data.rhs, data.lhs);
            break;
        }
        case InstInt: {
            auto data = std::get<DataInt>(inst.data);
            printf("Int %lld\n", data.value);
            break;
        }
        }
    }
}

void print_func(ref<Func> func) {
    printf("func %s (n %lld)\n", func->name.c_str(), func->params.size());
    print_block(func->root, 1);
}

void print_module(ref<Module> module) {
    for(auto func : module->funcs) {
        print_func(func);
    }
}